### PR TITLE
Re-implemented clustering of snapshots

### LINKED ISSUE #246: Implement clustering of snapshots
As outlined in the ML-DFT-Sandia - MALA merge.

### PR DESCRIPTION
**Pull request by RandomDefaultUser**
Closes #246 .

### CODE REVIEW

**Review by ellisja — CHANGES_REQUESTED** [review]
Very nice. Just a few small typos that I found.

File: mala/common/parameters.py [hunk 1]
```diff
@@ -401,6 +401,24 @@ class ParametersData(ParametersBase):
         If True, data is lazily loaded, i.e. only the snapshots that are
         currently needed will be kept in memory. This greatly reduces memory
         demands, but adds additional computational time.
+
+    use_clustering : bool
+        If True, and use_lazy_loading is True as well, the data is clustered,
+        i.e. not the entire training data is used by rather only a subset
+        which is determined by a clustering algorithm.
+
+    numbers_of_cluster : int
```
[the remark this answers is not in the document]
Comment (RandomDefaultUser) [reply mala/common/parameters.py]: Fixed in [0098144](https://github.com/mala-project/mala/pull/249/commits/0098144914ebd23b6979d060f9f2f8928eb0be41).

File: mala/common/parameters.py [hunk 2]
```diff
@@ -412,6 +430,10 @@ def __init__(self):
         self.input_rescaling_type = "None"
         self.output_rescaling_type = "None"
         self.use_lazy_loading = False
+        self.use_clustering = False
+        self.numbers_of_cluster = 0
+        self.train_ratio = 0.0
+        self.sample_ration = 0.0
```
[the remark this answers is not in the document]
Comment (RandomDefaultUser) [reply mala/common/parameters.py]: Fixed in [0098144](https://github.com/mala-project/mala/pull/249/commits/0098144914ebd23b6979d060f9f2f8928eb0be41).

File: mala/common/parameters.py [hunk 3]
```diff
@@ -401,6 +401,24 @@ class ParametersData(ParametersBase):
...
+    numbers_of_cluster : int
+        If use_clustering is True, this is the number of clusters used per
+        snapshot.
+
+    train_ratio : float
+        If use_clustering is True, this is the ratio of training data used
+        to train the encoder for the clustering.
+
+
+    sample_ration : float
```
[the remark this answers is not in the document]
Comment (RandomDefaultUser) [reply mala/common/parameters.py]: Fixed in [0098144](https://github.com/mala-project/mala/pull/249/commits/0098144914ebd23b6979d060f9f2f8928eb0be41).

**Comment by RandomDefaultUser** [comment]
Thanks for the feedback @ellisja ! I fixed the typos. I also deleted the files in the ml-dft-sandia folder that this merge request incorporates into the mala core code. One final question - in the implementation of this PR, only the training data set is clustered, validation and test data set are never clustered. Is that alright?

**Comment by ellisja** [comment]
> Thanks for the feedback @ellisja ! I fixed the typos. I also deleted the files in the ml-dft-sandia folder that this merge request incorporates into the mala core code. One final question - in the implementation of this PR, only the training data set is clustered, validation and test data set are never clustered. Is that alright?

Yes, that's right. The validation and test set are only used for forward inference calculations and would not benefit as much as the training set.

**Review by ellisja — APPROVED** [review]
